src/OpalCompiler-Core/OCUndeclaredVariableWarning.class.st

```diff
@@ -108,8 +108,10 @@ OCUndeclaredVariableWarning >> defineClass: className [
 		source: classDefinition;
 		logged: true;
 		evaluate.
-	^ (node owningScope lookupVar: className)
-		ifNil: [self error: 'should be not happen']
+	classBinding := node owningScope lookupVar: className.
+	"make sure to recompile all methods referencing this class"
+	classBinding usingMethods do: [:method | method recompile].
```

Comment:
It is unclear why it fixes any problem. The class definition evaluation (like in the browser) is supposed to recompile all required methods. Why it does not work in that case?

It looks like that the defining class not from the warning should lead to the same issue. Is it?

Reply:
If we define a class, the AST of the methods that reference that class still has the wrongly analysed variable, thinking
it is Undeclared (if the AST exists).

(the compiled methods are ok, the Undeclared mechanism made sure to update the bindings without the need to recompile).

For now, I just added a recompile step to make sure the methods and the AST (if there are some) are in sync… I wanted to avoid adding lots of code now in Pharo9, forcing a recompile should be an ok intermediate fix.

What we need to do for real is to instead fix the AST name analysis for this variable.

I will do that better version, but as part of Pharo10, I think, as part of the project to simplify handling of Undeclared Variables  (not doing it in the debugger or the Exception).

(I should have added more info to the issue tracker entry…)